community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityCacher.java

```diff
@@ -0,0 +1,146 @@
+    {
+        this.storeChannel = channel;
+        this.headerChannel = header;
+        this.previousGroupIds = new int[groupSlots];
+        for ( int i = 0; i < groupSlots; i++ )
+        {
+            previousGroupIds[i] = Group.GLOBAL.id();
+        }
+        // We don't really care about versions, it's just that apart from that the WritableLogChannel
+        // does precisely what we want and there's certainly value in not duplicating that functionality.
+        this.channel = new PhysicalWritableLogChannel(
+                new PhysicalLogVersionedStoreChannel( channel, 0, (byte) 0 ), bufferSize );
+        this.header = new PhysicalWritableLogChannel(
+                new PhysicalLogVersionedStoreChannel( header, 0, (byte) 0 ), 8*KB );
+    }
+
+    @Override
+    public void receive( ENTITY[] batch ) throws IOException
+    {
+        for ( ENTITY entity : batch )
+        {
+            writeEntity( entity );
+        }
+    }
+
+    protected void writeEntity( ENTITY entity ) throws IOException
+    {
+        // properties
+        if ( entity.hasFirstPropertyId() )
+        {
+            channel.putShort( (short) -1 ).putLong( entity.firstPropertyId() );
+        }
+        else
+        {
+            Object[] properties = entity.properties();
+            channel.putShort( safeCastLongToShort( properties.length/2 ) );
+            for ( int i = 0; i < properties.length; i++ )
+            {
+                String key = (String) properties[i++];
+                Object value = properties[i];
+                if ( value == null )
+                {
+                    continue;
+                }
+                writeToken( key );
+                writeValue( value );
+            }
+        }
+    }
+
+    protected void writeGroup( Group group, int slot ) throws IOException
+    {
+        if ( group.id() == previousGroupIds[slot] )
+        {
+            channel.put( (byte)0 );
+        }
+        else
+        {
+            channel.put( (byte)1 );
+            channel.putInt( previousGroupIds[slot] = group.id() );
+            writeToken( group.name() );
+        }
+    }
+
+    protected void writeValue( Object value ) throws IOException
+    {
+        ValueType type = ValueType.typeOf( value );
+        channel.put( type.id() );
+        type.write( value, channel );
+    }
+
+    protected void writeToken( String key ) throws IOException
+    {
+        Short id = tokens.get( key );
+        if ( id == null )
+        {
+            tokens.put( key, id = nextKeyId++ );
+            header.put( (byte) 1 );
+            ValueType.stringType().write( key, header );
+        }
+        channel.putShort( id );
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        header.put( (byte) 0 );
+        // This is a special value denoting the end of the stream. This is done like this since
+        // properties are the first thing read for every entity.
+        channel.putShort( (short) -2 );
```

Comment:
I think it is better to introduce a constant for this value and use it also in `InputEntityReader`